### PR TITLE
adding jdbc layer to the project

### PR DESCRIPTION
**Pull request by milena-hakobyan**
Screenshot 2025-06-20 at 16 27 35

### ✅ Summary

Implemented a JDBC support system for the project, including a utility class, test coverage using an embedded database, and a basic DAO layer.

---

### 🔧 **JDBC Utility Class**

- **Encapsulates connection handling** and simplifies query execution via prepared statements.
- **Methods implemented:**
  - `void execute(String query, Object... args)`  
    Executes basic DML operations (e.g., `INSERT`, `CREATE TABLE`) using prepared statements.
  - `void execute(String query, Consumer<PreparedStatement>)`  
    Exposes `PreparedStatement` to provide flexibility for more complex queries.
  - `<T> T findOne(String query, Function<ResultSet, T> mapper, Object... args)`  
    Fetches a **single result**, throws if more than one row is returned.
  - `<T> List<T> findMany(String query, Function<ResultSet, T> mapper, Object... args)`  
    Fetches **multiple results** into a list.

---

### 📚 **Comparison & Documentation**

- Compared both `execute` overloads in terms of **safety**, **flexibility**, and **use cases**.
- Added a Markdown file documenting (inside the resources folder):
  - **Trade-offs** between the two `execute` approaches.
  - **Pros and cons** of `DataSource` vs `DriverManager`, including:
    - Usage scenarios
    - Performance implications
    - Scalability differences

---

### 🧪 **Testing with Embedded H2**

- Integration tests use **H2 in-memory database**.
- **Preloaded schema and sample data** via SQL init scripts.
- Tests cover:
  - ✅ **Positive cases** (expected results)
  - ❌ **Negative cases** (error handling, edge conditions)

---

### 🗃️ **DAO Layer Implementation**

- Created a **basic data access layer** for domain entities (in the Repository folder).
- Supports **CRUD operations**.
- Included example for a **One-To-Many relationship**, such as fetching a parent with all its children.

### CODE REVIEW

File: src/main/java/com/example/repository/JdbcCourseRepo.java [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+                            u.last_login, u.is_active, s.progress_percentage, s.completed_courses, s.current_courses
+                     FROM enrollments e
+                     JOIN users u ON u.id = e.user_id
+                     JOIN students s ON s.user_id = u.id
+                     WHERE e.course_id = ?
+                """;
+
+        return dbConnection.findMany(query, rs -> {
+            try {
+                return new Student(rs.getInt("user_id"), rs.getString("username"), rs.getString("first_name"), rs.getString("last_name"), rs.getString("email"), rs.getString("password_hash"), rs.getTimestamp("last_login") != null ? rs.getTimestamp("last_login").toLocalDateTime() : null, rs.getBoolean("is_active"), rs.getDouble("progress_percentage"), rs.getInt("completed_courses"), rs.getInt("current_courses"));
```
Comment (kamirovgd) [inline src/main/java/com/example/repository/JdbcCourseRepo.java:87]: I'd extract it into it's own map* method for readability

File: src/main/java/com/example/repository/JdbcCourseRepo.java [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+
+    @Override
+    public void ensureCourseExists(Integer courseId) {
+        Objects.requireNonNull(courseId, "Course ID cannot be null");
+        if (findById(courseId).isEmpty()) {
+            throw new IllegalArgumentException("Course with ID '" + courseId + "' does not exist.");
+        }
+    }
+
+    private Course mapCourse(java.sql.ResultSet rs) {
```
Comment (kamirovgd) [inline src/main/java/com/example/repository/JdbcCourseRepo.java:125]: why do you need to specify the full package name here?

File: src/main/java/com/example/repository/JdbcGradeRepo.java [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+                    SELECT g.id AS grade_id, g.score, g.submission_id, g.feedback, g.graded_at,
+                           a.id, a.title, a.description, a.due_date, a.course_id
+                    FROM grades g
+                        JOIN submissions s ON g.submission_id = s.id
+                        JOIN assignments a ON s.assignment_id = a.id
+                    WHERE s.student_id = ? AND a.course_id = ?
+                """;
+
+        return dbConnection.findMany(query, rs -> {
+                    try {
```
Comment (kamirovgd) [inline src/main/java/com/example/repository/JdbcGradeRepo.java:104]: better to extract this

File: src/main/java/com/example/service/CourseService.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+    void createCourse(Course course);
+
+    void updateCourse(Course course);
+
+    void deleteCourse(Integer courseId);
+
+    Optional<Course> getCourseById(Integer courseId);
+
+    // Example of fetching a parent entity of One-To-Many relationship with all its children
```
Comment (kamirovgd) [inline src/main/java/com/example/service/CourseService.java:18]: does it avoid N+1 problem?

File: src/main/java/com/example/service/CourseService.java [hunk 5]
```diff
@@ -0,0 +1,58 @@
+package com.example.service;
+
+import com.example.model.*;
+
+import java.util.List;
+import java.util.Optional;
+
+public interface CourseService {
```
Comment (kamirovgd) [inline src/main/java/com/example/service/CourseService.java:8]: this interface violates Single Responsibility
it manages too many things

Comment (kamirovgd) [reply src/main/java/com/example/service/CourseService.java:8]: it's better to split to several separate interfaces

File: src/main/java/com/example/service/InstructorService.java [hunk 6]
```diff
@@ -0,0 +1,42 @@
+package com.example.service;
+
+import com.example.model.*;
+
+import java.util.List;
+import java.util.Optional;
+
+public interface InstructorService {
```
Comment (kamirovgd) [inline src/main/java/com/example/service/InstructorService.java:8]: this one too

File: src/main/java/com/example/service/UserService.java [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+public interface UserService {
+    Optional<User> getUserById(Integer id);
+
+    Optional<User> getUserByEmail(String email);
+
+    Optional<User> getUserByUserName(String userName);
+
+    List<User> getUsersByRole(Role role);
+
+    Student registerStudent(String username, String firstName, String lastName, String email, String rawPassword);
```
Comment (kamirovgd) [inline src/main/java/com/example/service/UserService.java:17]: this probably should be moved

File: src/main/java/com/example/utils/DatabaseConnection.java [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+
+    public Connection getConnection() {
+        try {
+            return dataSource.getConnection();
+        } catch (SQLException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    public void execute(String query, Object... args) {//args are the arguments to pass to the prepared statement
```
Comment (kamirovgd) [inline src/main/java/com/example/utils/DatabaseConnection.java:28]: no need in the comment

File: pom.xml [hunk 9]
```diff
@@ -37,11 +37,40 @@
...
+            <groupId>com.h2database</groupId>
+            <artifactId>h2</artifactId>
+            <version>2.3.232</version>
+            <scope>test</scope>
+        </dependency>
+
+        <dependency>
+            <groupId>com.h2database</groupId>
+            <artifactId>h2</artifactId>
+            <version>2.3.232</version> <!-- or latest available -->
```
[the remark this answers is not in the document]
Comment (milena-hakobyan) [reply pom.xml:50]: removed

File: pom.xml [hunk 10]
```diff
@@ -37,11 +37,40 @@
...
+            <scope>runtime</scope>
+        </dependency>
+
+        <!-- https://mvnrepository.com/artifact/com.zaxxer/HikariCP -->
```
[the remark this answers is not in the document]
Comment (milena-hakobyan) [reply pom.xml:54]: removed

File: src/main/java/com/example/repository/AnnouncementRepository.java [hunk 11]
```diff
@@ -0,0 +1,12 @@
+package com.example.repository;
+
+import com.example.model.Announcement;
+
+import java.util.List;
+
+public interface AnnouncementRepository extends CrudRepository<Announcement, Integer> {
+
+    List<Announcement> findByCourseId(Integer courseId);
```
[the remark this answers is not in the document]
Comment (milena-hakobyan) [reply src/main/java/com/example/repository/AnnouncementRepository.java:9]: done

File: src/main/java/com/example/repository/CrudRepository.java [hunk 12]
```diff
@@ -1,10 +1,12 @@
...
+package com.example.repository;
 
 import java.util.List;
 import java.util.Optional;
 
 public interface CrudRepository<T, ID> {
-    void save(T entity);
+    T save(T entity);
+
+    void update(T entity);
```
[the remark this answers is not in the document]
Comment (milena-hakobyan) [reply src/main/java/com/example/repository/CrudRepository.java:9]: changed the return type, accordingly modified the jdbcRepos

File: src/main/java/com/example/repository/JdbcAssignmentRepo.java [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+    }
+
+    @Override
+    public Optional<Assignment> findByTitle(String title) {
+        String query = "SELECT * FROM assignments WHERE title = ?;";
+        return Optional.ofNullable(dbConnection.findOne(query, this::mapAssignment, title));
+    }
+
+    @Override
+    public void addMaterial(Integer assignmentId, Material material) {
```
[the remark this answers is not in the document]
Comment (milena-hakobyan) [reply src/main/java/com/example/repository/JdbcAssignmentRepo.java:94]: changed it to be transactional

File: src/main/java/com/example/repository/JdbcCourseRepo.java (hunk 1 above)
Comment (milena-hakobyan) [reply src/main/java/com/example/repository/JdbcCourseRepo.java:87]: okay

File: src/main/java/com/example/repository/JdbcGradeRepo.java (hunk 3 above)
Comment (milena-hakobyan) [reply src/main/java/com/example/repository/JdbcGradeRepo.java:104]: extracted!

File: src/main/java/com/example/service/CourseService.java (hunk 4 above)
Comment (milena-hakobyan) [reply src/main/java/com/example/service/CourseService.java:18]: since it's for a single course, all lessons are fetched in one query, so the N+1 problem is avoided. right? 🤔

File: src/main/java/com/example/service/InstructorService.java (hunk 6 above)
Comment (milena-hakobyan) [reply src/main/java/com/example/service/InstructorService.java:8]: is there anything to modify now?

File: src/main/java/com/example/service/UserService.java (hunk 7 above)
Comment (milena-hakobyan) [reply src/main/java/com/example/service/UserService.java:17]: created a RegistrationService and moved associated methods there

File: src/main/java/com/example/service/CourseService.java (hunk 4 above)
Comment (kamirovgd) [reply src/main/java/com/example/service/CourseService.java]: yes, if everything is fetched in one go then it's ok

**Review by kamirovgd — APPROVED** [review]

File: src/main/java/com/example/service/CourseService.java (hunk 5 above)
Comment (kamirovgd) [reply src/main/java/com/example/service/CourseService.java]: fine for now
